File: lab3/jacobi.py
import numpy as np


def is_correct(m, a, x):
    rows, cols = m.shape
    for i in range(rows):
        s = sum(m[i, j] * x[j] for j in range(cols))
        if abs(s - a[i]) > exp: return False
    return True


if __name__ == '__main__':
    b = np.array([[32, 2, 1], [2, 32, 3], [1, 3, 6]], np.float32)
    c = np.array([2, 32, 0], np.float32)
    eps = 0.01

    rows, cols = b.shape
    x = np.zeros(cols, np.float32)
    temp_x = x.copy()

    i = 0
    while not is_correct(b, c, x) and i < 10:
        for j in range(rows):
            s = sum(b[j, k] * x[k] for k in range(cols) if k != j)
            temp_x[j] = (c[j] - s) / b[j, j]
        x = temp_x.copy()
        i += 1
        print(i, x)
    
    # checking

    results = np.zeros(cols, np.float32)

    for i in range(cols):
        results[i] = sum(b[i, j] * x[j] for j in range(cols))
    results = np.around(results)
    print(all(results == c))

File: lab3/Jacobi.hs
b = [[32, 2, 1], [2, 32, 3], [1, 3, 6]]
c = [2, 32, 0]
x = replicate (length b) 0
eps = 0.01


jacobi 100 _ = []
jacobi i x
    | not isCorrect = temp_x : jacobi (i + 1) temp_x
    | otherwise          = []
    where len = length b
          temp_x = getX 
          isCorrect = foldl (\acc x -> if and [ abs (s i - c !! i) < eps | i <- [0..len-1]] then True else acc) False x
              where s i = sum [b !! i !! j * x !! j | j <- [0..len-1]]
          getX = [(c !! j - s j) / b !! j !! j | j <- [0..len-1]] 
              where s j = sum [b !! j !! k * x !! k | k <- [0..len-1], k /= j]

main = putStrLn . unlines $ [(show $ i + 1) ++ ". " ++ (show $ val !! i) | i <- [0..len-1]]
   where len = length val
         val = jacobi 0 x